lib/Spawn.js

```diff
@@ -52,7 +52,18 @@ function Spawn(stream, command, args, name, output) {
 
   process.on('exit', function(code) {
     if (exited === false) {
-      spawnedProcess.kill();
+      // 'exit' callback cannot perform asynchronous work
+      // therefore run tree-kill in a subprocess using spawnSync.
+
+      // Resolve the path to tree-kill as we cannot assume our
+      // subprocess will have the same module resolution as this package
+      // due to the way node_modules is nested by package managers
+      var treeKillModulePath = require.resolve('tree-kill');
```

Comment:
Have you tested simply calling `treeKill()` without waiting for the callback to complete?
For example, would it be as simple as doing:
```javascript
var treeKill = require('tree-kill');

...

if (exited === false) {
  // is the command sent and are processes killed without waiting for a callback?
  treeKill(spawnedProcess.pid);
}
```

Reply:
Yes, you can see in the first of my three commits I tried this here https://github.com/arjunmehta/multiview/pull/12/commits/f8be961fa744ab22e8f0cc38789265686ebfee12
The node process stops as soon it returns to the event loop for the first time so the `treeKill()` doesn't get to do any of its work  (I believe tree-kill begins with some async information gathering before it even gets to killing anything).

Reply:
Ah gotcha. The other thing to consider is that the `tree-kill` module has a "binary" cli tool included with the module.

So you should be able to simply do:
```
childProcess.spawnSync('tree-kill', [spawnedProcess.pid])
```
I'd like to remove the complexity of needing to resolve the module path if possible, or find a more elegant way of doing so. Hopefully the binary executable would do this. Also, I'm not sure we'd need to handle output from this process (ie. no need to inherit `stdio`).

Please let me know your thoughts

Reply:
Regarding the tree-kill binary, I considered that approach, however I don't believe that the `node_modules` system guarantees that `bin`s of a sub-dependency would be placed in a specific location.
I did a test and while yarn and my relatively-up-to-date npm version seems to like to hoist everything up to the root `node_modules`, npm still exhibits the classical non-hoisting behaviour when you install modules globally. See this screenshot:
![image](https://user-images.githubusercontent.com/1409112/59707747-1abb4000-91b8-11e9-9522-d01f4c887305.png)

This means we'd need to "resolve" the path to the 'tree-kill' binary similar or identical to how we resolve its js for the purpose of launching node.

For this reason, I think invoking the js entry point directly, with the node executable that is already running, resolved using the module system's resolution mechanism, is the most robust approach with the least assumptions / complications.

I chose to inherit stdio so any errors emitted by `tree-kill` aren't completely silent. I'm not sure multiview should itself fail when the tree-kill has an issue, but at least they get reported.

LMK if you disagree with any of this.

Reply:
Ahh. Seems messy and unpredictable indeed. I'm just feeling like the solution to spawn command to run a module is not very clean and seems counter intuitive.

Looking into this more, I'm wondering if the [`beforeExit`](https://nodejs.org/api/process.html#process_event_beforeexit) event is actually more suitable for this. It can handle asynchronous code:

```javascript
process.on('beforeExit', function(code) {
  if (exited === false) {
    treeKill(spawnedProcess.pid, function() {
      process.exit(code);
    });
  }
});
```

Reply:
I completely sympathise with your desire to find a cleaner/simpler approach here.

I also considered 'beforeExit' however it's also not that simple. You see, beforeExit is not called when the default sigint handler runs. So you might think, let's install a sigint handler in multiview. However I consider it to be quite unfriendly to install signal handlers in any library, because installing a signal handler modifies global process state, replacing the default handler that an app may have assumed would run, and furthermore multiple signal handlers in a node process can get tricky because if you don't process.exit(), you're not acting like the default handler and so you may have prevented shutdown, but if you DO process.exit() you may prevent other sigint handlers installed by the outer application from running.
For these reasons I will strongly avoid registering a signal handler in a library, and instead ask the calling application to inform the library of signals / invoke shutdown explicitly. This is feasible, but complicates multiview's API slightly and increases the chance of not using multiview correctly as a library, which is why I looked for some other approach.

If multiview was purely a cli tool I'd have definitely used signal handlers + 'beforeExit'.

We can probably switch the CLI and the library to use this signal / cleanup call approach if you prefer it to the spawnSync.

A wise Dev once told me "If it looks silly but it works well, it's not silly", which I like to think could apply here.

Reply:
Hah, I think you've clearly thought it through and it seems you know the best technical way to resolve this. I have to admit I'm mostly concerned with the implementation, which I think can be worked out.

So, I have one more suggestion (which I may just do at a later time), but would love to hear your thoughts: Put the tree-kill script (require, deriving module path etc) you're writing into a separate module that can be run in its own process. You may need to use path.resolve somewhere to find the right path.
```
var treeKill = require('tree-kill');

var processID = process.argv[2];
treeKill(processID, function(err) {
  if(err) console.error(err);
})
```

Then the behavior can be localized and the code much cleaner and maintainable. No need to use require.resolve, or eval or build a script string to evaluate in the main Spawn.js:
```
spawnSync(process.execPath, ['./our-own-custom-treekill.js', spawnedProcess.pid], { stdio: 'inherit' });
```

Again, thank you for bearing with this and for your work.

Reply:
I think splitting it out in to our own wrapper is reasonable and makes adding more code to the subprocess easier for sure, but it doesn't avoid module resolution challenges entirely.

When you run `spawnSync(process.execPath, ['./standalone-treekill.js', ...`, the cwd of our process (and the one inherited by the child) wont be the folder that contains `standalone-treekill.js` so you couldn't use the path `'./standalone-treekill.js'`. However we can maybe use `require.resolve()` to get the path OR if you're wanting to avoid `require.resolve()` we can maybe use `__dirname` or `__filename` (and `import.meta.url` on newer node versions) to get the path to that script, then we can either set the child CWD or pass the full script path in the first param.

Thoughts?

Reply:
If tree-kill is a dependency of our project, `multiview/node_modules/tree-kill/index.js` exists.

We will need to use `__dirname` and `path.resolve` but node will always go by the order specified [here](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders) to load the right module. It will load the module require'd from the main multiview `node_modules` location.

```javascript
// assuming tree-kill in same directory as Spawn.js
var treeKillScriptPath = path.resolve(__dirname, './standalone-treekill.js');
spawnSync(process.execPath, [treeKillScriptPath, spawnedProcess.pid], { stdio: 'inherit' });
```

No need to `import.meta.url`. Executing the script with the right path should load the right dependency always.

Reply:
I realized now we're saying the same thing! Yes, I assumed we'd need to resolve the `dirname` and execute that. 👍

Reply:
So if multiview is being used as a library and it's `my-app/node_modules/multiview/node_modules/tree-kill`, and the cwd is currently `my-app`. You're saying let's always run with the cwd `multiview` and then let node get tree-kill out of its `node_modules`? That sounds like it works.

Reply:
It’ll find the nearest one up the parent tree and should find it because it’ll be a dependency of multiview. In a way, it should resolve the same as if we were to require it in Spawn.js.